File: algorithms/Sorting/Hard/MergeSort/VSDMergeSort.java
/*Copyrights to vsdevelopers.io*/
/*For more programs visit vsdevelopers.io */
/*Java program for Merge sort*/


public class Main {
	
	
	static void VSDmerge(int[] array,int l, int r)
	{
		if(l<r) {
			int m =l+(r-l)/2;
			//Recursively calling merge method to divide the array
			VSDmerge(array,l,m);
			VSDmerge(array,m+1,r);
			//Combines the divided array pieces in sorted order
			VSDMergesort(array,l,m,r);
		}
	}
	 
   static void VSDMergesort(int[] array,int l, int m, int r){
	   //Assigning sizes for temporary arrays
	   	int n1 = m-l+1;
        int n2 = r-m;
        //Temporary arrays to sort divided elements
        int Left[] = new int[n1];
        int Right[] = new int[n2];
        //Assigning elements
        for (int i=0;i<n1;i++)
            Left[i] = array[l + i];
        for (int j=0;j<n2;j++)
            Right[j] = array[m+1+j];
        //Initial indexes to sort elements using temporary arrays
        int i = 0,j = 0;
        //Index for original array
        int k = l;
        //Comparing elements to arrange them in sorted order in original array
        while (i<n1 && j<n2) {
            if (Left[i]<=Right[j]) {
                array[k]=Left[i];
                i++;
            }
            else {
                array[k]=Right[j];
                j++;
            }
            k++;
        }
        //Arranging remaining elements from left array
        while (i < n1) {
           array[k]=Left[i];
            i++;
            k++;
        }
      //Arranging remaining elements from right array
        while (j < n2) {
            array[k]=Right[j];
            j++;
            k++;
        }
        
    }
   //Function to take user input

//Function to print output
public static void VSDprintOutput(int[] array) {
	System.out.println("The sorted order is:");
	for(int a:array){
        System.out.print(a+" ");
    }
    System.out.println();
}
public static void main(String args[]) {
	int[] array={1, 2, 53, 12, 45, 11, 56};
    int size=array.length;
	VSDmerge(array,0,size-1);//Calling merge for the entire array
	VSDprintOutput(array);//Printing output
}
}

File: algorithms/Sorting/Hard/MergeSort/VSDMergeSort.cpp

#include<iostream>

using namespace std;


void VSDmerge(int arr[],int l,int m,int r){

    int n1=m-l+1;
    int n2=r-m;

    int left[n1];
    int right[n2];

    for(int i=0;i<n1;i++){
        left[i]=arr[l+i];
    }
    for(int j=0;j<n2;j++){
        right[j]=arr[m+1+j];
    }
    int i=0,j=0;
    int k=l;
    while(i<n1 && j<n2){
        if(left[i]<=right[j]){
            arr[k]=left[i];
            i++;
        }
        else{
            arr[k]=right[j];
            j++;
        }
        k++;
    }
    while(i<n1){
        arr[k]=left[i];
        i++;
        k++;
    }

    while(j<n2){
        arr[k]=right[j];
        j++;
        k++;
    }
    

}
void VSDmergesort(int arr[],int l,int r){
    if(l<r){
        int m=l+(r-l)/2;
        VSDmergesort(arr,l,m);
        VSDmergesort(arr,m+1,r);
        VSDmerge(arr,l,m,r);
    }
}
int main(){
    int n=7;
    int arr[]={2,24,12,96, 456, 899,34};
    VSDmergesort(arr,0,n-1);
    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";

    }
    cout<<endl;
    return 0;
}